### PR TITLE
Create BeatTable #238 #239 #240

### LINKED ISSUE #238: BeatTable.getPair
Behaviour: Supporting functionality/BeatTable;the pair related to a and b can be obtained

@tested_feature("Supporting functionality")

@tested_operation("BeatTable")

@tested_behaviour("the pair related to a and b can be obtained")

by getPair(Choice a, Choice b)

[Deviation in model](http://adadocs.demokracia.rulez.org/PDEngine/edemo/master/index.html#de4b570a-17b0-4872-a614-22acfca5eea2-b46d19bb-27bf-46bf-9096-cea0caec2ca8-96b3aa1a-d34b-4c9b-bde2-643835f5ee75)

### PR DESCRIPTION
**Pull request by valentinbujdoso**
Fix #238 #239 #240

### CODE REVIEW

**Comment by magwas** [comment]
The beat table is supposed to contain pairs. See https://github.com/andrewcmyers/civs/blob/master/cgi-bin/beatpath2.pm
Entities describe data structures, hence contain only public fields, and live in package org.rulez.demokracia.pdengine.dataobjects
The behaviour related to an entity is in a class derived from it.
Use Matrix as the backing type for a beat table.
Listen to sonar. A lot of times he is right.
If you have questions, ask. I'm here to help.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+		
+		assertThrows(() -> beatTable.beatInformation(null, null, Direction.DIRECTION_FORWARD)
+				).assertMessageIs("Invalid row key");
+	}
+	
+	@TestedFeature("Supporting functionality")
+	@TestedOperation("BeatTable")
+	@TestedBehaviour("the beat information related to a and b can be obtained for forward and backward")
+	@Test
+	public void beatTable_with_empty_matrix() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java]: I would say this tests the case when the key does not belong to the table.
Please name the test accordingly.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		
+		assertThrows(() -> beatTable.beatInformation(choice1, choice2, Direction.DIRECTION_FORWARD)
+				).assertMessageIs("Invalid row key");
+	}
+	
+	@TestedFeature("Supporting functionality")
+	@TestedOperation("BeatTable")
+	@TestedBehaviour("the beat information related to a and b can be obtained for forward and backward")
+	@Test
+	public void beatTable_forward() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java]: google translate gives and acceptable translation for "beatInformation forward irányban megadja choice1 choice2-vel szembeni nyeréseinek számát "

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		beatTable.setPair(choice1, choice2, pair);
+		
+		assertEquals(pair.winning, beatTable.beatInformation(choice1, choice2, Direction.DIRECTION_FORWARD));
+	}
+	
+	@TestedFeature("Supporting functionality")
+	@TestedOperation("BeatTable")
+	@TestedBehaviour("the beat information related to a and b can be obtained for forward and backward")
+	@Test
+	public void beatTable_backward() throws IsNullException {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java]: same in English

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import org.rulez.demokracia.pdengine.testhelpers.CreatedDefaultChoice;
+
+public class BeatTableGetPairTest extends CreatedDefaultChoice {
+
+	@Before
+	public void setUp() throws ReportedException {
+		super.setUp();
+	}
+
+	@TestedFeature("Supporting functionality")
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: please move up the @Tested* annotation to the class

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+				).assertMessageIs("Invalid row key");
+	}
+	
+	@TestedFeature("Supporting functionality")
+	@TestedOperation("BeatTable")
+	@TestedBehaviour("the pair related to a and b can be obtained")
+	@Test
+	public void getPair_with_not_proper_inputs() {
+		ArrayList<Choice> list = new ArrayList<Choice>();
+		Choice choice1 = new Choice("name1", "userName1");
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: violation of TDD rule #2 http://butunclebob.com/ArticleS.UncleBob.TheThreeRulesOfTdd

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+		assertThrows(() -> beatTable.getPair(choice1, choice3)
+				).assertMessageIs("Invalid row key");
+	}
+	
+	@TestedFeature("Supporting functionality")
+	@TestedOperation("BeatTable")
+	@TestedBehaviour("the pair related to a and b can be obtained")
+	@Test
+	public void getPair_with_empty_matrix() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: see comment for the similar test for the other method

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import org.rulez.demokracia.pdengine.BeatTable.Direction;
+import org.rulez.demokracia.pdengine.annotations.TestedBehaviour;
+import org.rulez.demokracia.pdengine.annotations.TestedFeature;
+import org.rulez.demokracia.pdengine.annotations.TestedOperation;
+import org.rulez.demokracia.pdengine.dataobjects.Pair;
+import org.rulez.demokracia.pdengine.exception.IsNullException;
+import org.rulez.demokracia.pdengine.exception.ReportedException;
+import org.rulez.demokracia.pdengine.testhelpers.CreatedDefaultChoice;
+
+public class BeatTableGetPairTest extends CreatedDefaultChoice {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: I doubt we need that many test to introduce functionality of calling a single function with two arguments.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableSetPairTest.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+import org.rulez.demokracia.pdengine.BeatTable.Direction;
+import org.rulez.demokracia.pdengine.annotations.TestedBehaviour;
+import org.rulez.demokracia.pdengine.annotations.TestedFeature;
+import org.rulez.demokracia.pdengine.annotations.TestedOperation;
+import org.rulez.demokracia.pdengine.dataobjects.Pair;
+import org.rulez.demokracia.pdengine.exception.IsNullException;
+import org.rulez.demokracia.pdengine.exception.ReportedException;
+import org.rulez.demokracia.pdengine.testhelpers.CreatedDefaultChoice;
+
+public class BeatTableSetPairTest extends CreatedDefaultChoice {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableSetPairTest.java]: all the comments for getPair test apply

**Review by magwas — CHANGES_REQUESTED** [review]
Also I expect sonar to find legitimate issues, please take care of them, too.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java (hunk 4 above)
Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: So something like this:
	@TestedFeature("Supporting functionality")
	@TestedOperation("BeatTable")
	@TestedBehaviour("the pair related to a and b can be obtained")
public class BeatTableGetPairTest extends CreatedDefaultChoice {

Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: ?

File: src/main/java/org/rulez/demokracia/pdengine/dataobjects/Pair.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.rulez.demokracia.pdengine.persistence.BaseEntity;
+
+public class Pair extends BaseEntity {
+	
+	private static final long serialVersionUID = 1L;
+	public int winning;
+	public int losing;
+	
+	public Pair() {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/dataobjects/Pair.java]: This is not covered. Do we need it?

File: src/main/java/org/rulez/demokracia/pdengine/BeatTable.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+		if (beat1 == null || beat2 == null)
+			throw new IllegalArgumentException("Invalid Pair key");
+		
+		if(beat1.winning > beat2.winning)
+			return beat1;
+		else if (beat1.winning == beat2.winning)
+			if(beat1.losing < beat2.losing)
+				return beat1;
+			else if(beat1.losing == beat2.losing) 
+				throw new IllegalArgumentException("Can not decide");
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: A tie is a legitimate result in the business logic. Do not throw up on it.

File: src/main/java/org/rulez/demokracia/pdengine/BeatTable.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+			else
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: I feel like the next four lines can be shortened to just

    return beat2;

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	@Test
+	public void The_1st_key_should_be_a_Choice() {
+		BeatTable beatTable = new BeatTable();
+		
+		assertThrows(() -> beatTable.beatInformation(null, null, Direction.DIRECTION_FORWARD)
+				).assertMessageIs("Invalid row key");
+	}
+	
+	@Test
+	public void beatTable_when_the_key_does_not_belong_to_the_table() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableBeatInformationTest.java]: Please use a legit English sentence.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	@TestedBehaviour("if beats tie, looses decide")
+	@Test
+	public void compareBeats_with_equal_forward_and_lower_beat1_backward() {
+		result = beatTable.compareBeats(beats1, beats3);
+		assertTrue(result.winning == beats1.winning && result.losing == beats1.losing);	
+	}
+	
+	@TestedBehaviour("if beats are different, the bigger wins")
+	@Test
+	public void compareBeats_with_equal_forward_and_equal_backward_beats() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java]: a tie is a legit result.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+@TestedBehaviour("the pair related to a and b can be obtained")
+public class BeatTableGetPairTest extends CreatedDefaultChoice {
+
+	@Before
+	public void setUp() throws ReportedException {
+		super.setUp();
+	}
+
+	@Test
+	public void getPair_with_not_defined_inputs() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableGetPairTest.java]: a correct English sentence, pleas

**Review by magwas — CHANGES_REQUESTED** [review]
At least some of the code duplication findings, none of the linguistic antipattern and all other findings by sonar seem to be legit. Of course do not comment, but fix scope.

File: src/main/java/org/rulez/demokracia/pdengine/BeatTable.java (hunk 10 above)
Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: But which one is the "winner" in this case?

Comment (magwas) [reply src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: None. This is a third case.
In the meantime I have realized that we can approach this case by deferring design choice.
Let's treat this case the same way as we would treat a mocked up method: use the unimplemented testcase markups for the relevant test, and let the ones implementing cases based on a tie decide what should be returned here.

**Comment by magwas** [comment]
there are still legit sonar findings.

**Comment by valentinbujdoso** [comment]
@magwas As I see there are some duplicated codes in sonar but I cannot eliminate it in a nice way.
Could you take a look?

**Comment by magwas** [comment]
There are actual code duplications using.

		result = beatTable.matrix.getElement(choice1, choice2);
		assertTrue(result.winning == pair.winning && result.losing == pair.losing);

Also you use the exact same code to test different things in setPair and getPair test, which means that some of the tests are redundant.

Please note that there are other problems, for example the complexity of BeatTable.
I have experimented with the traits pattern we were talking about, and BeatTable can be made as simple as
```
public class BeatTable extends MapMatrix<Choice, Pair> implements Matrix<Choice, Pair>, ContainingBeats {

	private static final long serialVersionUID = 1L;

	public BeatTable(final Collection<Choice> keyCollection) {
		super(keyCollection);
	}

}
```

File: src/main/java/org/rulez/demokracia/pdengine/BeatTable.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+	public BeatTable() {
+		this(new ArrayList<Choice>());
+	}
+
+	public BeatTable(final Collection<Choice> keyCollection) {
+		super(keyCollection);
+	}
+
+	public int beatInformation(final Choice choice1, final Choice choice2, final Direction direction) {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: Factor out everything which is not getter/setter or constructor to a treat (interface with default methods.

File: src/main/java/org/rulez/demokracia/pdengine/BeatTable.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+		if (direction.equals(Direction.DIRECTION_FORWARD))
+			result = pair.winning;
+		else
+			result = pair.losing;
+
+		return result;
+	}
+
+	public Pair getPair(final Choice choice1, final Choice choice2) {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: It is not a good idea to shadow a getter and setter just to delegate to the normal getter/setter.
Please unimplement them.
Yeah, I know, this was in the design. But the design is not good.

File: src/main/java/org/rulez/demokracia/pdengine/dataobjects/Pair.java [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.rulez.demokracia.pdengine.persistence.BaseEntity;
+
+public class Pair extends BaseEntity {
+	
+	private static final long serialVersionUID = 1L;
+	public int winning;
+	public int losing;
+	
+	public Pair(final int winning, final int losing) {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/dataobjects/Pair.java:7]: not even constructor in an entity.

**Review by magwas — CHANGES_REQUESTED** [review]

File: src/main/java/org/rulez/demokracia/pdengine/BeatTable.java (hunk 16 above)
Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: If I unimplement the getPair how can I get the pairs in the interface?

Comment (magwas) [reply src/main/java/org/rulez/demokracia/pdengine/BeatTable.java]: with getElement

**Comment by valentinbujdoso** [comment]
@magwas Could you run the sonar to not whine about the public entites?

File: src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
+package org.rulez.demokracia.pdengine;
+
+import org.rulez.demokracia.pdengine.BeatTable.Direction;
+import org.rulez.demokracia.pdengine.dataobjects.Pair;
+import org.rulez.demokracia.pdengine.exception.ReportedException;
+import org.rulez.demokracia.types.Matrix;
+
+public interface ContainingBeats extends Matrix<Choice, Pair>{
+	
+	long serialVersionUID = 1L;
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java:10]: is this needed here, in an interface?

File: src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+	default void checkPair(final Pair pair) {
+		if (pair == null)
+			throw new ReportedException("Invalid Pair key");
+	}
+	
+	default int beatInformation(final Choice choice1, final Choice choice2, final Direction direction) {
+		if (direction == null)
+			throw new IllegalArgumentException("Invalid direction");
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java]: We are throwing ReportedErrors

File: src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+			return beat1;
+		else if (beat1.winning < beat2.winning)
+			return beat2;
+		else
+			if (beat1.losing < beat2.losing)
+				return beat1;
+			else if (beat2.losing < beat1.losing)
+				return beat2;
+			else
+				throw new IllegalArgumentException("Can not decide");
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java]: You mean ReportedError?

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+public class BeatTableCompareBeatsTest extends CreatedBeatTable {
+
+	@Before
+	public void setUp() throws ReportedException {
+		super.setUp();
+	}
+
+	@TestedBehaviour("if beats are different, the bigger wins")
+	@Test
+	public void compareBeats_throws_an_exception_when_first_input_param_is_not_defined() {
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java]: somehow the first and second are mixed inthis and the next test

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+	@Test
+	public void compareBeats_gives_back_the_backward_lower_beat1() {
+		setCompareBeatsAsResult(beats1, beats3);	
+		assertTrue(result.winning == beats1.winning && result.losing == beats1.losing);
+	}
+	
+    @TestedBehaviour("Unimplemented")
+	@Test
+	public void compareBeats_is_not_yet_implemented_when_the_result_is_tie() {
+		assertThrows(() -> beatTable.compareBeats(beats1, beats1)
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java]: use assertUnimplemented

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+@TestedFeature("Schulze method")
+@TestedOperation("compare beats")
+public class BeatTableCompareBeatsTest extends CreatedBeatTable {
+
+	@Before
+	public void setUp() throws ReportedException {
+		super.setUp();
+	}
+
+	@TestedBehaviour("if beats are different, the bigger wins")
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java]: You should give this behaviour its own test class, as it is tested by more than 3 tests, and this makes PMD nervous.

**Review by magwas — CHANGES_REQUESTED** [review]

File: src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java (hunk 18 above)
Comment (valentinbujdoso) [reply src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java:10]: If the interface does not contain this id the Ci build fail. So i dont know what to do. I know this is not the nice solution.

Comment (magwas) [reply src/main/java/org/rulez/demokracia/pdengine/ContainingBeats.java:10]: strange... then leave it here.

File: src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java (hunk 22 above)
Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java]: as I see there is no such a method like assertUnimplemented yet.

Comment (valentinbujdoso) [reply src/test/java/org/rulez/demokracia/pdengine/BeatTableCompareBeatsTest.java]: Ok I think I created that method I hope you will like it. :)

**Review by magwas — APPROVED** [review]
Nice work

**Comment by magwas** [comment]
testcaseCount=10

**Pull request merged**